src/main/java/org/jahia/modules/sitemap/filter/SitemapCacheFilter.java

```diff
@@ -59,126 +54,85 @@
+            return "";
         }
 
-        // re-render by returning null; manually flush jahia cache prior to render
-        CacheHelper.flushOutputCachesForPath(resource.getPath(), false);
         return null;
     }
 
     /**
-     * Create/refresh file cache with rendered contents (previousOut) if needed.
+     * Create cache, if needed, with rendered contents (previousOut) and expiration time.
      */
     @Override
-    public String execute(String previousOut, RenderContext renderContext, Resource resource, RenderChain chain)
-            throws RepositoryException
-    {
-        if (!needsCaching(resource)) return previousOut;
-
-        JCRNodeWrapper sitemapNode = resource.getNode();
-        JCRNodeWrapper cacheNode = getCacheNode(sitemapNode);
-
-        InputStream inputStream = null;
-        try {
-            // refresh if invalid cache; cache only if there's contents
-            if (!isValidCache(cacheNode) && StringUtils.isNotBlank(previousOut)) {
-                inputStream = IOUtils.toInputStream(previousOut, StandardCharsets.UTF_8);
-                refreshSitemapCache(sitemapNode, inputStream);
-            }
-        } finally {
-            IOUtils.closeQuietly(inputStream);
-        }
-
-        return previousOut;
-    }
+    public String execute(
+            String previousOut,
+            RenderContext renderContext,
+            Resource resource,
+            RenderChain chain
+    ) throws RepositoryException {
 
-    /**
-     * Create or refresh file cache for given parent for a given locale
-     * Uses parent session to save
-     * https://stackoverflow.com/questions/4685959/get-file-out-of-jcr-file-node
-     */
-    private JCRNodeWrapper refreshSitemapCache(JCRNodeWrapper sitemapNode, InputStream data) throws RepositoryException {
-        String cacheName = getCacheName(sitemapNode);
-        String nodePath = sitemapNode.getPath();
-
-        // use system session to be able to save file cache in live workspace
-        return JCRTemplate.getInstance().doExecuteWithSystemSessionAsUser(null,
-                Constants.LIVE_WORKSPACE, null, new JCRCallback<JCRNodeWrapper>() {
-            @Override public JCRNodeWrapper doInJCR(JCRSessionWrapper session) throws RepositoryException {
-                JCRNodeWrapper node = session.getNode(nodePath);
-                node.uploadFile(cacheName, data, "application/xml");
-                session.save();
-                return node;
-            }
-        });
-    }
+        if (needsCaching(resource)) return previousOut;
 
+        // if we are here, the cache does not exist or has expired.
 
-    /** Apply file caching only for default template */
-    public boolean needsCaching(Resource resource) {
-        String templateName = resource.getTemplate();
-        return "lang".equalsIgnoreCase(templateName);
-    }
+        String targetSitemapCacheKey = PREFIX_CACHE_NAME + resource.getNode().getLanguage();
```

Comment:
The key is not taking in account the node path that can change in case of:
- different site
- different sitemap in the same site
Here all sitemap for a given language will be the same

Reply:
So true I had not paid attention to this detail, sorry, fixed